### PR TITLE
Add support for Django 3.0

### PR DESCRIPTION
**Pull request by adamf**
This primarily changes the use of deprecated imports, and adds
some logic for retaining support for python2.7 imports where needed.

Note that JSONField and multiselect, which are 3rd party modules,
do not yet support django 3 and both need updating for this to work:
JSONField needs the changes in imports, and multiselect needs a
method signature to be updated. I'll send pull requests to those
maintainers.

### CODE REVIEW

**Comment by adamf** [comment]
Note that https://github.com/adamchainz/django-jsonfield/pull/18 (not by me) will add Django 3.0 support to `django-jsonfield`; that said, it might be worth creating an issue on this project to move off `django-jsonfield` since the maintainer doesn't do active development on it anymore.

**Comment by adamf** [comment]
Bumping the multiselect version in `requirements-test.txt` to `0.1.10` will fix the django 3 testing issues, but breaks `py27-django-111`. Not sure how many folks you have using this that use Django 1.11.

File: src/auditlog/middleware.py [hunk 1]
```diff
@@ -5,7 +5,10 @@
 
 from django.conf import settings
 from django.db.models.signals import pre_save
-from django.utils.functional import curry
+try:
+    from functools import partialmethod
```
Comment (adamf) [inline src/auditlog/middleware.py]:
```suggestion
    from functools import partial
```

File: src/auditlog/middleware.py [hunk 2]
```diff
@@ -43,7 +46,7 @@ def process_request(self, request):
 
         # Connect signal for automatic logging
         if hasattr(request, 'user') and is_authenticated(request.user):
-            set_actor = curry(self.set_actor, user=request.user, signal_duid=threadlocal.auditlog['signal_duid'])
+            set_actor = partialmethod(self.set_actor, user=request.user, signal_duid=threadlocal.auditlog['signal_duid'])
```
Comment (adamf) [inline src/auditlog/middleware.py]:
```suggestion
            set_actor = partial(self.set_actor, user=request.user, signal_duid=threadlocal.auditlog['signal_duid'])
```

**Comment by emanuelmd** [comment]
Hey, is there any plan to merge this?

At the company I work we're using this package in one of our projects and we'd like to upgrade to Django 3.0. Happy to help with anything.

Best

**Comment by adamf** [comment]
Friendly ping. I'd love to see this merged.

**Comment by sssppp57** [comment]
This constantly breaks my CI build.  Can you merge it please?

**Comment by aviyadeveloper** [comment]
Just mentioning - I'm going to have to pass on using this application due to this issue. This is blocking all python 3 systems from even trying out the application.

**Comment by l0b0** [comment]
Has anyone had any luck migrating from django-auditlog to a similar module? I'd be very interested to hear your findings, as I'm on a project which will soon have to either use a fork or migrate away from django-auditlog.

@jjkester If this project is no longer maintained, could you please add a note to that effect in the readme?

**Comment by aviyadeveloper** [comment]
> Has anyone had any luck migrating from django-auditlog to a similar module? I'd be very interested to hear your findings, as I'm on a project which will soon have to either use a fork or migrate away from django-auditlog.
> 
> @jjkester If this project is no longer maintained, could you please add a note to that effect in the readme?

I ended up writing something on my own with the use of Django's built-in Log Entry functionality and it works quite decently.

File: src/auditlog/middleware.py [hunk 3]
```diff
@@ -5,7 +5,10 @@
 
 from django.conf import settings
 from django.db.models.signals import pre_save
-from django.utils.functional import curry
+try:
+    from functools import partial
+except ImportError:
+    from django.utils.functional import curry as partialmethod
```
Comment (pcamelo) [inline src/auditlog/middleware.py:11]: `partialmethod` is not used in the code, should it be `partial` instead ?

**Comment by pcamelo** [comment]
@adamf could you look into my comment above, this code have issues with old python versions, it just needs a small fix.

File: src/auditlog/middleware.py (hunk 3 above)
Comment (paulpavlishcoupa) [reply src/auditlog/middleware.py:11]: Thanks for the check on this @pcamelo. I see that this [fork made the change](https://github.com/vinaygode/django-auditlog/commit/cc87b00d60866f986bd21e993dfab18fbf1ab89d) from @vinaygode. 

I haven't tried it with Django 2.2, but that seems like right way to handle curry and partial. This is also recommended on [Django 3 release notes](https://docs.djangoproject.com/en/3.0/releases/3.0/).

**Comment by RobertAKARobin** [comment]
Can this be merged?

**Comment by paulpavlishcoupa** [comment]
> Can this be merged?

I think they need to make a slight modification based on @pcamelo's [comment](https://github.com/jjkester/django-auditlog/pull/227#discussion_r431107108), but I haven't heard from @pcamelo or @adamf.

A question for the maintainers of this project:
Does the PR also need line coverage improvement? I see that there is a lower percentage coverage according to codecov. Would you also like that increased before it is merged?

image

**Comment by pcamelo** [comment]
@paulpavlishcoupa's PR seems ok. From my side it is OK to go. @adamf could you merge @paulpavlishcoupa changes to your code and update the PR ?

**Pull request closed without merging**